### PR TITLE
Ng modules: Lifecycyle hooks and ApplicationRef cleanup

### PR DESCRIPTION
**Pull request by tbosch**

### CODE REVIEW

File: modules/@angular/compiler/src/provider_analyzer.ts [hunk 1]
```diff
@@ -478,10 +478,13 @@ function _resolveProviders(
...
-      const lifecycleHooks = provider.useClass ? provider.useClass.lifecycleHooks : [];
+      const lifecycleHooks =
+          provider.token.identifier && provider.token.identifier instanceof CompileTypeMetadata ?
+          provider.token.identifier.lifecycleHooks :
+          [];
       resolvedProvider = new ProviderAst(
-          provider.token, provider.multi, eager, [provider], providerType, lifecycleHooks,
-          sourceSpan);
+          provider.token, provider.multi, eager || lifecycleHooks.length > 0, [provider],
+          providerType, lifecycleHooks, sourceSpan);
```
Comment (IgorMinar) [inline modules/@angular/compiler/src/provider_analyzer.ts]: all providers with hooks eager

File: modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts [hunk 2]
```diff
@@ -77,12 +77,11 @@ export function bindDirectiveAfterViewLifecycleCallbacks(
   }
 }
 
-export function bindDirectiveDestroyLifecycleCallbacks(
-    directiveMeta: CompileDirectiveMetadata, directiveInstance: o.Expression,
-    compileElement: CompileElement) {
+export function bindInjectableDestroyLifecycleCallbacks(
+    provider: ProviderAst, directiveInstance: o.Expression, compileElement: CompileElement) {
```
Comment (IgorMinar) [inline modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts]: directiveInstance => providerInstance

File: modules/@angular/core/src/linker/ng_module_factory.ts [hunk 3]
```diff
@@ -87,4 +102,18 @@ export abstract class NgModuleInjector<T> extends CodegenComponentFactoryResolve
   get injector(): Injector { return this; }
 
   get componentFactoryResolver(): ComponentFactoryResolver { return this; }
+
+  destroy(): void {
+    if (this._destroyed) {
+      throw new BaseException(
+          `This module is already destroyed (${stringify(this.instance.constructor)})`);
```
Comment (IgorMinar) [inline modules/@angular/core/src/linker/ng_module_factory.ts]: The ng modules <name> has been already destroyed.

File: modules/@angular/core/src/application_ref.ts [hunk 4]
```diff
@@ -238,21 +247,37 @@ export abstract class PlatformRef {
...
    */
   abstract dispose(): void;
 
+  /**
+   * Destroy the Angular platform and all Angular applications on the page.
+   */
+  abstract destroy(): void;
+
+  /**
+   * @deprecated Use `destroyd` instead
```
Comment (IgorMinar) [inline modules/@angular/core/src/application_ref.ts]: typo: destroyed

File: modules/@angular/core/src/application_ref.ts [hunk 5]
```diff
@@ -277,29 +302,41 @@ function _callAndReportToExceptionHandler(
...
+  get disposed() { return this.destroyed; }
+  get destroyed() { return this._destroyed; }
 
-  dispose(): void {
-    ListWrapper.clone(this._applications).forEach((app) => app.dispose());
-    this._disposeListeners.forEach((dispose) => dispose());
-    this._disposed = true;
+  destroy() {
+    if (this._destroyed) {
+      throw new BaseException('The platform is already destroyed!');
```
Comment (IgorMinar) [inline modules/@angular/core/src/application_ref.ts]: ... has already been destroyed

File: modules/@angular/router-deprecated/src/router.ts [hunk 6]
```diff
@@ -548,6 +548,11 @@ export class RootRouter extends Router {
     return promise;
   }
 
+  /**
+   * @internal
```
Comment (IgorMinar) [inline modules/@angular/router-deprecated/src/router.ts]: can't be internal

File: modules/@angular/router/src/router.ts [hunk 7]
```diff
@@ -197,6 +197,11 @@ export class Router {
   }
 
   /**
+   * @internal
```
Comment (IgorMinar) [inline modules/@angular/router/src/router.ts]: can't be internal

File: modules/@angular/core/src/application_tokens.ts [hunk 8]
```diff
@@ -54,6 +54,13 @@ export const PLATFORM_INITIALIZER: any = new OpaqueToken('Platform Initializer')
 export const APP_INITIALIZER: any = new OpaqueToken('Application Initializer');
 
 /**
+ * All callbacks provided via this token will be called when a component has been bootstrapped.
```
Comment (IgorMinar) [inline modules/@angular/core/src/application_tokens.ts]: add info:
- add signature
- info that callbacks are expected to be synchronous
- callbacks will be called multiple times if multiple components are bootstrapped

File: modules/@angular/core/test/linker/ng_module_integration_spec.ts [hunk 9]
```diff
@@ -310,7 +311,7 @@ function declareTests({useJit}: {useJit: boolean}) {
             .toBe(SomeComp);
       });
 
-      it('should entryComponents ComponentFactories in imported modules', () => {
+      it('should crate ComponentFactories in imported modules', () => {
```
Comment (IgorMinar) [inline modules/@angular/core/test/linker/ng_module_integration_spec.ts]: create

File: modules/@angular/core/src/application_init.ts [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+ */
+export const APP_INITIALIZER: any = new OpaqueToken('Application Initializer');
+
+/**
+ * A class that reflects the state of running {@link APP_INITIALIZER}s.
+ *
+ * @experimental
+ */
+@Injectable()
+export class AppInitStatus {
```
Comment (IgorMinar) [inline modules/@angular/core/src/application_init.ts]: let's check if this shouldn't be ApplicationInitStatus to make the apis consistent

File: modules/@angular/core/src/application_init.ts [hunk 11]
```diff
@@ -0,0 +1,49 @@
...
+import {isPromise} from '../src/facade/lang';
+
+import {Inject, Injectable, OpaqueToken, Optional} from './di';
+
+
+/**
+ * A function that will be executed when an application is initialized.
+ * @experimental
+ */
+export const APP_INITIALIZER: any = new OpaqueToken('Application Initializer');
```
Comment (IgorMinar) [inline modules/@angular/core/src/application_init.ts]: APPLICATION_INITIALIZER?

File: tools/public_api_guard/core/index.d.ts [hunk 12]
```diff
@@ -119,11 +119,21 @@ export declare abstract class AnimationWithStepsMetadata extends AnimationMetada
...
+export declare const APP_BOOTSTRAP_LISTENER: OpaqueToken;
+
+/** @experimental */
 export declare const APP_ID: any;
 
 /** @experimental */
 export declare const APP_INITIALIZER: any;
 
+/** @experimental */
+export declare class AppInitStatus {
```
Comment (IgorMinar) [inline tools/public_api_guard/core/index.d.ts]: ApplicationInitStatus

File: modules/@angular/core/src/linker/ng_module_factory.ts (hunk 3 above)
Comment (tbosch) [inline modules/@angular/core/src/linker/ng_module_factory.ts]: Done

File: modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts (hunk 2 above)
Comment (tbosch) [inline modules/@angular/compiler/src/view_compiler/lifecycle_binder.ts]: Done

File: modules/@angular/core/src/application_ref.ts (hunk 4 above)
Comment (tbosch) [inline modules/@angular/core/src/application_ref.ts]: Done

File: modules/@angular/core/src/application_ref.ts (hunk 5 above)
Comment (tbosch) [inline modules/@angular/core/src/application_ref.ts]: Done

File: modules/@angular/router-deprecated/src/router.ts (hunk 6 above)
Comment (tbosch) [inline modules/@angular/router-deprecated/src/router.ts]: Done

File: modules/@angular/router/src/router.ts (hunk 7 above)
Comment (tbosch) [inline modules/@angular/router/src/router.ts]: Done

File: modules/@angular/core/src/application_tokens.ts (hunk 8 above)
Comment (tbosch) [inline modules/@angular/core/src/application_tokens.ts]: Done

File: modules/@angular/core/test/linker/ng_module_integration_spec.ts (hunk 9 above)
Comment (tbosch) [inline modules/@angular/core/test/linker/ng_module_integration_spec.ts]: Done

File: modules/@angular/core/src/application_init.ts (hunk 10 above)
Comment (tbosch) [inline modules/@angular/core/src/application_init.ts]: Done

File: modules/@angular/core/src/application_init.ts (hunk 11 above)
Comment (tbosch) [inline modules/@angular/core/src/application_init.ts]: Not changing it in this PR... I just moved it...

File: tools/public_api_guard/core/index.d.ts (hunk 12 above)
Comment (tbosch) [inline tools/public_api_guard/core/index.d.ts]: Done

**Comment by tbosch** [comment]
Reviewed in person with @IgorMinar

**Pull request closed without merging**